### DISCUSSION TITLE
If (not) startswith mutually exclusive steps

### DISCUSSION THREAD

**Discussion by valentijnscholten** [community/community#26386] [Actions] [answered]
<p>Hi,</p>
<p>I am trying to have two mutually exclusive steps in my workflow, based on an input parameter’s value:</p>
<pre><code class="lang-auto">      - name: Create release branch
        if: startsWith(${{ github.event.inputs.from_branch }}, 'release/') != true
        run: |
          echo "NEW_BRANCH=release/${{ github.event.inputs.release_number }}" &gt;&gt; $GITHUB_ENV
      - name: Use existing release branch
        if: startsWith(${{ github.event.inputs.from_branch }}, 'release/')
        run: |
          echo "NEW_BRANCH=${{ github.event.inputs.from_branch }}" &gt;&gt; $GITHUB_ENV
</code></pre>
<p>But if I specify ‘dev’ as input, both steps are executed. If I specify ‘release/0.0.0’ as input, also both steps get executed. Whatever input I give, always both steps are executed. As if the if statements / startswith don’t work at all.</p>
<p>I tried also:</p>
<pre><code class="lang-auto">!(startsWith(${{ github.event.inputs.from_branch }}, 'release/'))
"!startsWith(${{ github.event.inputs.from_branch }}, 'release/')"
"!(startsWith(${{ github.event.inputs.from_branch }}, 'release/'))"
</code></pre>
<p>None of them seem to work. I read <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions" rel="noopener nofollow ugc">https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions</a>, but there’s no example of how to use the ‘not’ operator.</p>
<p>Valentijn</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/valentijnscholten">@valentijnscholten</a>,</p>
<p>You are using the incorrect expression syntax for the <strong>if</strong> conditional in your workflow.<br>
Change them like as this can work:</p>
<pre><code class="lang-yaml">- name: Create release branch
  if: ${{ !startsWith(github.event.inputs.from_branch, 'release/') }}
  run: |
    echo "NEW_BRANCH=release/${{ github.event.inputs.release_number }}" &gt;&gt; $GITHUB_ENV

- name: Use existing release branch
  if: ${{ startsWith(github.event.inputs.from_branch, 'release/') }}
  run: |
    echo "NEW_BRANCH=${{ github.event.inputs.from_branch }}" &gt;&gt; $GITHUB_ENV
</code></pre>
<p>There are few points you need to know when setting <strong>if</strong> conditionals (<a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idif" rel="noopener nofollow ugc"><strong>jobs.&lt;job_id&gt;.if</strong></a> or <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsif" rel="noopener nofollow ugc"><strong>jobs.&lt;job_id&gt;.steps.if</strong></a>) in your workflow:</p>
<ol>
<li>
<p>If you want to use the expression syntax “<strong>${{ }}</strong>” in the <strong>if</strong> conditionals, you should use it to surround the whole condition statement, and you do not need to use it for each referenced context expression in the condition statement.</p>
<ul>
<li>
<p>The correct syntax.</p>
<pre><code class="lang-yaml">if: ${{ startsWith(github.event.inputs.from_branch, 'release/') }}  ✅
</code></pre>
</li>
<li>
<p>The incorrect syntax.</p>
<pre><code class="lang-yaml">if: startsWith(${{ github.event.inputs.from_branch }}, 'release/')  ❌
</code></pre>
</li>
</ul>
</li>
<li>
<p>In most cases, you can directly omit the expression syntax “<strong>${{ }}</strong>” in the <strong>if</strong> conditionals.</p>
<pre><code class="lang-yaml">if: startsWith(github.event.inputs.from_branch, 'release/')  ✅
</code></pre>
</li>
<li>
<p>If the condition statement is started with “<strong>!</strong>” (<strong>Not</strong>), the expression syntax “<strong>${{ }}</strong>” can’t be omitted.</p>
<ul>
<li>The correct syntax.<pre><code class="lang-yaml">if: ${{ !startsWith(github.event.inputs.from_branch, 'release/') }}  ✅
</code></pre>
</li>
<li>The incorrect syntax.<pre><code class="lang-yaml">if: !startsWith(github.event.inputs.from_branch, 'release/')  ❌
</code></pre>
</li>
</ul>
</li>
</ol>
<p>In addition, you can <a href="https://docs.github.com/en/free-pro-team@latest/actions/managing-workflow-runs/enabling-debug-logging" rel="noopener nofollow ugc">enable debug logging</a> to view the result of the <strong>if</strong> conditionals for each step from the debug logs. From the logs, you can see how the job evaluates the <strong>if</strong> conditionals for each step and returns the final result (<strong>true</strong> or <strong>false</strong>).<br>
For example:</p>
<ul>
<li><a href="https://github.com/TestWorkflowsRML/Event-Triggers/runs/1362472094?check_suite_focus=true#step:4:2" rel="noopener nofollow ugc">https://github.com/TestWorkflowsRML/Event-Triggers/runs/1362472094?check_suite_focus=true#step:4:2</a></li>
<li><a href="https://github.com/TestWorkflowsRML/Event-Triggers/runs/1362472094?check_suite_focus=true#step:5:2" rel="noopener nofollow ugc">https://github.com/TestWorkflowsRML/Event-Triggers/runs/1362472094?check_suite_focus=true#step:5:2</a></li>
</ul>

**Comment by valentijnscholten** [comment]
<p>Wow, best reply ever. Thank you <a class="mention" href="/brightran">@brightran</a> !</p>

**Comment by Simran-B** [comment]
<aside class="quote group-github-partners" data-username="brightran" data-post="2" data-topic="141841">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<p>If the condition statement is started with “ <strong>!</strong> ” ( <strong>Not</strong> ), the expression syntax “ <strong>${{ }}</strong> ” can’t be omitted.</p>
</blockquote>
</aside>
<p>That’s really good to know. Are there other cases which require <code>${{ }}</code>? And are they listed in the documentation? They really should be!</p>

**Comment by Simran-B** [comment]
<p>I suppose this has less to do with GitHub Actions and more with YAML in general.</p>
<ul>
<li>A leading <code>!</code> is reserved for tagging</li>
<li>Leading <code>*</code> and <code>&amp;</code> are reserved for anchors</li>
<li>A leading question mark followed by whitespace is reserved for complex mapping keys:<pre><code class="lang-yaml">? [foo, bar]
: baz
</code></pre>
</li>
</ul>
<p>With the if condition wrapped in <code>${{ }}</code>, issues with above special characters are avoided because the string starts with <code>$</code>. Alternatively, one can also write:</p>
<ul>
<li>
<pre><code class="lang-yaml">if: &gt;-
  !startsWith(github.event.inputs.from_branch, 'release/')
</code></pre>
</li>
<li><code>if: "!startsWith(github.event.inputs.from_branch, 'release/')"</code></li>
<li><code>if: '!startsWith(github.event.inputs.from_branch, ''release/'')'</code></li>
</ul>

**Comment by funes79** [comment]
why is this not in the official [documentation](https://docs.github.com/en/github-ae@latest/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idif)? 👍 @BrightRan

**Reply by JoshWeepie** [reply]
Dude yes I was literally trying to do this today, this doc is lacking sorely in examples.

https://docs.github.com/en/actions/learn-github-actions/expressions#startswith
